lib/concurrent/executor/safe_task_executor.rb

```diff
@@ -6,17 +6,15 @@ module Concurrent
   # success - indicating if the callable has been executed without errors
   # value - filled by the callable result if it has been executed without errors, nil otherwise
   # reason - the error risen by the callable if it has been executed with errors, nil otherwise
-  class SafeTaskExecutor
+  class SafeTaskExecutor < Synchronization::Object
 
     def initialize(task, opts = {})
-      @task = task
-      @mutex = Mutex.new
-      @exception_class = opts.fetch(:rescue_exception, false) ? Exception : StandardError
+      super(task, opts)
     end
```

Comment:
@ianks, @jdantonio: the `@Mutex` i's one part of it. (It's just my convention though so feel free to ignore.) Second part is that `ensure_ivar_visibility!` has to be called after the final fields are set to make them visible after initialization of the object. There is plenty of examples how it can be used in `edge/future.rb`.